src/engines/bing.rs

```diff
@@ -0,0 +1,109 @@
+//! The `bing` module handles the scraping of results from the bing search engine
+//! by querying the upstream bing search engine with user provided query and with a page
+//! number if provided.
+
+use std::collections::HashMap;
+
+use regex::Regex;
+use reqwest::header::HeaderMap;
+use reqwest::Client;
+use scraper::Html;
+
+use crate::models::aggregation_models::SearchResult;
+
+use crate::models::engine_models::{EngineError, SearchEngine};
+
+use error_stack::{Report, Result, ResultExt};
+
+use super::search_result_parser::SearchResultParser;
+
+/// A new Bing engine type defined in-order to implement the `SearchEngine` trait which allows to
+/// reduce code duplication as well as allows to create vector of different search engines easily.
+pub struct Bing {
+    /// The parser, used to interpret the search result.
+    parser: SearchResultParser,
+}
+
+impl Bing {
+    /// Creates the Bing parser.
+    pub fn new() -> Result<Self, EngineError> {
+        Ok(Self {
+            parser: SearchResultParser::new(
+                ".b_results",
+                ".b_algo",
+                "h2 a",
+                ".tpcn a.tilk",
+                ".b_caption p",
+            )?,
+        })
+    }
+}
+
+#[async_trait::async_trait]
+impl SearchEngine for Bing {
+    async fn results(
+        &self,
+        query: &str,
+        page: u32,
+        user_agent: &str,
+        client: &Client,
+        _safe_search: u8,
+    ) -> Result<HashMap<String, SearchResult>, EngineError> {
+        // Bing uses `start results from this number` convention
+        // So, for 10 results per page, page 0 starts at 1, page 1
+        // starts at 11, and so on.
+        let results_per_page = 10;
+        let start_result = results_per_page * page + 1;
+
+        let url: String = match page {
+            0 => {
+                format!("https://www.bing.com/search?q={query}")
+            }
+            _ => {
+                format!("https://www.bing.com/search?q={query}&first={start_result}")
+            }
+        };
+
+        let header_map = HeaderMap::try_from(&HashMap::from([
+            ("USER_AGENT".to_string(), user_agent.to_string()),
+            ("REFERER".to_string(), "https://google.com/".to_string()),
+            (
+                "CONTENT_TYPE".to_string(),
+                "application/x-www-form-urlencoded".to_string(),
+            ),
+        ]))
+        .change_context(EngineError::UnexpectedError)?;
```

Comment:
I think I would suggest adding a `cookie` header because this helps prevent being detected as a bot by the upstream search engine. 

Also, the cookie for bing will look something like the following:
```
MUID=2D0BD206B856600B0802C1DBB95F61B9; _EDGE_V=1; MUIDB=2D0BD206B856600B0802C1DBB95F61B9; SRCHD=AF=NOFORM; SRCHUID=V=2&GUID=B68AA9D86D6F45C5A891D99D8902C941&dmnchg=1; _HPVN=CS=eyJQbiI6eyJDbiI6MiwiU3QiOjAsIlFzIjowLCJQcm9kIjoiUCJ9LCJTYyI6eyJDbiI6MiwiU3QiOjAsIlFzIjowLCJQcm9kIjoiSCJ9LCJReiI6eyJDbiI6MiwiU3QiOjAsIlFzIjowLCJQcm9kIjoiVCJ9LCJBcCI6dHJ1ZSwiTXV0ZSI6dHJ1ZSwiTGFkIjoiMjAyMy0xMi0zMVQwMDowMDowMFoiLCJJb3RkIjowLCJHd2IiOjAsIlRucyI6MCwiRGZ0IjpudWxsLCJNdnMiOjAsIkZsdCI6MCwiSW1wIjo0LCJUb2JuIjowfQ==; _Rwho=u=d; ai_session=TB676+0gO13S3BW5AYVW3d|1704013243633|1704013243633; BCP=AD=0&AL=0&SM=0; ipv6=hit=1704016843140&t=4; _EDGE_S=SID=13B4DE4C147D60EB12DECDB415A661CB&mkt=en-kw; USRLOC=HS=1&ELOC=LAT=29.333396911621094|LON=47.99052810668945|N=Mirqab%2C%20Al%20Asimah|ELT=1|; SRCHUSR=DOB=20231231&T=1704013237000&TPC=1704013248000; bngps=s=0; _UR=QS=0&TQS=0; _RwBf=r=0&ilt=10&ihpd=1&ispd=6&rc=18&rb=0&gb=0&rg=200&pc=15&mtu=0&rbb=0&g=0&cid=&clo=0&v=10&l=2023-12-31T08:00:00.0000000Z&lft=0001-01-01T00:00:00.0000000&aof=0&ard=0001-01-01T00:00:00.0000000&o=2&p=&c=&t=0&s=0001-01-01T00:00:00.0000000+00:00&ts=2023-12-31T09:04:39.0642186+00:00&rwred=0&wls=&wlb=&wle=&ccp=&lka=0&lkt=0&aad=0&TH=; _SS=SID=13B4DE4C147D60EB12DECDB415A661CB&R=18&RB=0&GB=0&RG=200&RP=15; SRCHHPGUSR=SRCHLANG=en&IG=5AFEB9E3084249A9AF2DADF971C5965F&PV=6.6.8&BRW=NOTP&BRH=M&CW=819&CH=703&SCW=1164&SCH=3480&DPR=1.0&UTC=180&DM=1&HV=1704013480&WTS=63839610037&PRVCW=1325&PRVCH=703&NEWWND=0&NRSLT=10c&LSL=0&AS=1&NNT=0&HAP=2&VSRO=0&CHTRSP=3&WEBTHEME=1
```

Reply:
I did not include the cookies because their cookies are very long and generated per-user. How about we include the shorter ones like `_EDGE_v=1`, `SRCHD=AF=NOFORM`, etc, what do you think?

Reply:
> I did not include the cookies because their cookies are very long and generated per-user. How about we include the shorter ones like `_EDGE_v=1`, `SRCHD=AF=NOFORM`, etc, what do you think?

Ok I see, I think then I would suggest adding them and checking if everything works still. Also, if there are more cookie parameters then I would suggest exploring them and adding them too. :slightly_smiling_face: